warehouse/accounts/views.py

```diff
@@ -322,6 +323,50 @@ def webauthn_authentication_validate(request):
     return {"fail": {"errors": errors}}
 
 
+@view_config(
+    route_name="accounts.recovery-code",
+    renderer="accounts/recovery-code.html",
+    uses_session=True,
+    require_csrf=True,
+    require_methods=False,
+    has_translations=True,
+)
+def recovery_code(request, _form_class=RecoveryCodeAuthenticationForm):
+    if request.authenticated_userid is not None:
+        return HTTPSeeOther(request.route_path("manage.projects"))
+
+    try:
+        two_factor_data = _get_two_factor_data(request)
+    except TokenException:
+        request.session.flash(_("Invalid or expired two factor login."), queue="error")
+        return HTTPSeeOther(request.route_path("accounts.login"))
+
+    userid = two_factor_data.get("userid")
+    redirect_to = two_factor_data.get("redirect_to")
+
+    user_service = request.find_service(IUserService, context=None)
+
+    form = _form_class(**request.POST, user_id=userid, user_service=user_service)
+
+    if request.method == "POST":
+        if form.validate():
+            _login_user(request, userid, two_factor_method="recovery-code")
+
+            resp = HTTPSeeOther(redirect_to)
+            resp.set_cookie(
+                USER_ID_INSECURE_COOKIE,
+                hashlib.blake2b(str(userid).encode("ascii"), person=b"warehouse.userid")
+                .hexdigest()
+                .lower(),
+            )
```

Comment:
It would be super tidy if we could bundle all of this up in to the `_login_user` function. It might be a bit messy though. If you have a tidier suggestion for how to incorporate that refactor into this PR I'm happy to take a crack at it, but it's probably better handled in a separate changeset.

Reply:
Sounds good 👍